File: src/main/java/com/enderzombi102/cmt/zone/FunctionZone.java
package com.enderzombi102.cmt.zone;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.server.function.CommandFunction;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.Vec3d;

import java.util.Optional;

import static com.enderzombi102.cmt.CustomMapsTools.LOGGER;

public class FunctionZone extends AbstractZone<PlayerEntity> {

	public final Identifier enterFunction;
	public final Identifier leaveFunction;
	public final Identifier midFunction;
	private CommandFunction enterFunc;
	private CommandFunction midFunc;
	private CommandFunction leaveFunc;

	public FunctionZone(ServerWorld world, Vec3d pos0, Vec3d pos1, String id, ZoneActionData data) {
		super(world, pos0, pos1, id, PlayerEntity.class );
		this.enterFunction = data.enterFunction;
		this.leaveFunction = data.leaveFunction;
		this.midFunction = data.midFunction;
		this.loadFunc();
	}

	// serialization
	public FunctionZone(ServerWorld world, NbtCompound tag) {
		super(world, tag, PlayerEntity.class);
		this.enterFunction = new Identifier( tag.getString("efunc") );
		this.midFunction = new Identifier( tag.getString("mfunc") );
		this.leaveFunction = new Identifier( tag.getString("lfunc") );
		this.loadFunc();
	}

	private void loadFunc() {
		// get possible function objects
		final Optional<CommandFunction> enter = this.server.getCommandFunctionManager().getFunction(this.enterFunction);
		final Optional<CommandFunction> stay = this.server.getCommandFunctionManager().getFunction(this.midFunction);
		final Optional<CommandFunction> leave = this.server.getCommandFunctionManager().getFunction(this.leaveFunction);

		// check if we found all objects + log if not
		if ( enter.isPresent() ) this.enterFunc = enter.get();
		else LOGGER.warn("[CMT] function <" + this.enterFunction + "> was not found");
		if ( stay.isPresent() ) this.midFunc = stay.get();
		else LOGGER.warn("[CMT] function <" + this.midFunction + "> was not found");
		if ( leave.isPresent() ) this.leaveFunc = leave.get();
		else LOGGER.warn("[CMT] function <" + this.leaveFunction + "> was not found");
	}

	@Override
	public void onEnter(PlayerEntity entity) {
		this.server.getCommandFunctionManager().execute(
				this.enterFunc,
				this.server.getCommandSource().withEntity(entity).withWorld(this.world)
		);
	}

	@Override
	public void onLeave(PlayerEntity entity) {
		this.server.getCommandFunctionManager().execute(
				this.leaveFunc,
				this.server.getCommandSource().withEntity(entity).withWorld(this.world)
		);
	}

	@Override
	public void onTick(PlayerEntity entity) {
		this.server.getCommandFunctionManager().execute(
				this.midFunc,
				this.server.getCommandSource().withEntity(entity).withWorld(this.world)
		);
	}

	// serialization
	@Override
	public void writeToNbt(NbtCompound tag) {
		super.writeToNbt(tag);
		tag.putString( "efunc", this.enterFunction.toString() );
		tag.putString( "mfunc", this.midFunction.toString() );
		tag.putString( "lfunc", this.leaveFunction.toString() );
	}
}

File: src/main/kotlin/com/enderzombi102/cmt/zone/FunctionZone.kt
package com.enderzombi102.cmt.zone

import com.enderzombi102.cmt.CustomMapsTools
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.NbtCompound
import net.minecraft.server.function.CommandFunction
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Identifier
import net.minecraft.util.math.Vec3d

class FunctionZone : AbstractZone<PlayerEntity> {
	val enterFunction: Identifier?
	val leaveFunction: Identifier?
	val midFunction: Identifier?
	private var enterFunc: CommandFunction? = null
	private var midFunc: CommandFunction? = null
	private var leaveFunc: CommandFunction? = null

	constructor(world: ServerWorld, pos0: Vec3d, pos1: Vec3d, id: String, data: ZoneActionData) : super(
		world,
		pos0,
		pos1,
		id,
		PlayerEntity::class.java
	) {
		enterFunction = data.enterFunction
		leaveFunction = data.leaveFunction
		midFunction = data.midFunction
		loadFunc()
	}

	// serialization
	constructor(world: ServerWorld?, tag: NbtCompound) : super(world, tag, PlayerEntity::class.java) {
		enterFunction = Identifier(tag.getString("efunc"))
		midFunction = Identifier(tag.getString("mfunc"))
		leaveFunction = Identifier(tag.getString("lfunc"))
		loadFunc()
	}

	private fun loadFunc() {
		// get possible function objects
		val enter = server.commandFunctionManager.getFunction(enterFunction)
		val stay = server.commandFunctionManager.getFunction(midFunction)
		val leave = server.commandFunctionManager.getFunction(leaveFunction)

		// check if we found all objects + log if not
		if (enter.isPresent) enterFunc =
			enter.get() else CustomMapsTools.LOGGER.warn("[CMT] function <" + enterFunction + "> was not found")
		if (stay.isPresent) midFunc =
			stay.get() else CustomMapsTools.LOGGER.warn("[CMT] function <" + midFunction + "> was not found")
		if (leave.isPresent) leaveFunc =
			leave.get() else CustomMapsTools.LOGGER.warn("[CMT] function <" + leaveFunction + "> was not found")
	}

	override fun onEnter(entity: PlayerEntity) {
		server.commandFunctionManager.execute(
			enterFunc,
			server.commandSource.withEntity(entity).withWorld(world)
		)
	}

	override fun onLeave(entity: PlayerEntity) {
		server.commandFunctionManager.execute(
			leaveFunc,
			server.commandSource.withEntity(entity).withWorld(world)
		)
	}

	override fun onTick(entity: PlayerEntity) {
		server.commandFunctionManager.execute(
			midFunc,
			server.commandSource.withEntity(entity).withWorld(world)
		)
	}

	// serialization
	override fun writeToNbt(tag: NbtCompound) {
		super.writeToNbt(tag)
		tag.putString("efunc", enterFunction.toString())
		tag.putString("mfunc", midFunction.toString())
		tag.putString("lfunc", leaveFunction.toString())
	}
}
